server/src/main/java/com/cloud/network/IpAddressManagerImpl.java

```diff
@@ -1425,6 +1416,32 @@ public IPAddressVO associateIPToGuestNetwork(long ipId, long networkId, boolean
         }
     }
 
+    /**
+     * Prevents associating an IP address to an allocated (unimplemented network) network, throws an Exception otherwise
+     * @param owner Used to check if the user belongs to the Network
+     * @param ipToAssoc IP address to be associated to a Network, can only be associated to an implemented network for Source NAT
+     * @param network Network to which IP address is to be associated with, must not be in allocated state for Source NAT Network/IP association
+     * @return true if IP address can be successfully associated with Source NAT network
+     */
+    protected boolean isSourceNatAvailableForNetwork(Account owner, IPAddressVO ipToAssoc, Network network) {
+        NetworkOffering offering = _networkOfferingDao.findById(network.getNetworkOfferingId());
+        boolean sharedSourceNat = offering.isSharedSourceNat();
+        boolean isSourceNat = false;
+        if (!sharedSourceNat) {
+            if (getExistingSourceNatInNetwork(owner.getId(), network.getId()) == null) {
+                if (network.getGuestType() == GuestType.Isolated && network.getVpcId() == null && !ipToAssoc.isPortable()) {
+                    if (network.getState() == Network.State.Allocated) {
+                        //prevent associating an ip address to an allocated (unimplemented network).
+                        //it will cause the ip to become source nat, and it can't be disassociated later on.
+                        throw new InvalidParameterValueException("Network is in allocated state, implement network first before acquiring an IP address");
```

Comment:
can you improve this message a little bit? I mean, could you add the network UUID in the message to inform, and register in the logs the network that generated this exception.

Reply:
I have added the network name to the message of the thrown Exception. I do not think Logging here is necessary since the exception will be caught by the entry code and rethrown.

Reply:
Ah sure, I do agree with you that logging is not necessary here.

What I meant is that if we add the UUID to the exception description, when it is logged (I guess the exception will be written to the log at some point), we will be able to have a better way to identity which network caused the error.

Reply:
Good point, I have included the UUID in the message of the Exception.